### PR TITLE
tests: boardfarm: port Client association link metrics test

### PR DESCRIPTION
**Pull request by abelog**
## Description

Port test_client_association_link_metrics from test_flows.py to
boardfarm.

## Changes

- Port `test_client_association_link_metrics` from `test_flows.py`
- Create new device to represent dummy station (`PrplMeshStationDummy`)
- Use new device in ported test
- Change naming scheme in boardfarm configuration file, so we can address devices directly by name
- Split dummy device docker image name and name of device.
In order to address devices by name, name set in configuration file should be a member of boardfarm build-in [enum](https://github.com/mattsm/boardfarm/blob/100521fde1fb67536682cafecc2f91a6e2e8a6f8/boardfarm/lib/DeviceManager.py#L25).  But name of docker images will be same, so different user's won't be able to run test at same time. To avoid it, split name of device and name of docker image spawned by the test.

## Notes

Those changes breaking all other tests, due to introduction of new device. As boardfarm isn't used at the moment it's not critical. Fix of other tests are introduced in #1486 PR, to ease review of this PR.
Part of: https://jira.prplfoundation.org/browse/PPM-12

### CODE REVIEW

File: tests/boardfarm_plugins/boardfarm_prplmesh/tests/client_association_link_metrics.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    an Associated STA Link Metrics Query message with an Associated STA Link Metrics
+    Response message containing an Associated STA Link Metrics TLV for the associated STA.'''
+
+    def runTest(self):
+        # Locate test participants in array
+        for dev in self.dev:
+            if dev.controller_entity:
+                controller = dev.controller_entity
+            if dev.agent_entity:
+                agent = dev.agent_entity
```
Comment (arnout) [inline tests/boardfarm_plugins/boardfarm_prplmesh/tests/client_association_link_metrics.py]: It's really time we refactor this setup and teardown work.

I'll convert this comment to an issue and assign it to @aciliketcap

File: tests/boardfarm_plugins/boardfarm_prplmesh/tests/client_association_link_metrics.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+        # Regression check
+        # Don't connect not existing STAtion
+        dev.wired_sniffer.start(self.__class__.__name__ + "-" + dev.name)
+        sta_mac = "11:11:33:44:55:66"
+        debug("Send link metrics query for unconnected STA")
+        controller.ucc_socket.dev_send_1905(agent.mac, 0x800D,
```
Comment (arnout) [inline tests/boardfarm_plugins/boardfarm_prplmesh/tests/client_association_link_metrics.py:39]: I think the ALEntity has a helper method so you don't need the explicit ucc_socket:
```suggestion
        controller.dev_send_1905(agent.mac, 0x800D,
```

File: tests/boardfarm_plugins/boardfarm_prplmesh/tests/client_association_link_metrics.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                                            tlv(0x95, 0x0006, '{sta_mac}'.format(sta_mac=sta_mac)))
+        self.check_log(agent,
+                       "client with mac address {sta_mac} not found".format(sta_mac=sta_mac),
+                       timeout=30)
```
Comment (arnout) [inline tests/boardfarm_plugins/boardfarm_prplmesh/tests/client_association_link_metrics.py:43]: Please explain in the commit message why you need such a huge timeout.

I guess it's because of the overhead of doing a shell command on the target - although I can't image that would take 30 seconds... Anyway, if that's the case, then the _check_log function that does the shell command should simply add a constant offset to the timeout.

File: tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py [hunk 4]
```diff
@@ -33,7 +33,8 @@ def __init__(self, *args, **kwargs):
         # Getting unic ID to distinguish devices and network they belong to
         self.unique_id = os.getenv("SUDO_USER", os.getenv("USER", ""))
 
-        self.name = "-".join((config.get("name", "prplmesh_docker"), self.unique_id))
+        self.name = config.get("name", "prplmesh_docker")
```
Comment (arnout) [inline tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py:36]: Is self.name something that is used by boardfarm?

If not, I think it would be better to call this something else, and to keep using self.name for the thing with unique_id appended. Otherwise it's confusing that PrplMeshDocker.name of "prplmesh_docker", but the corresponding ALEntity.name is "prplmesh_docker-arnout".

File: tests/boardfarm_plugins/boardfarm_prplmesh/prplmesh_config.json [hunk 5]
```diff
@@ -1,19 +1,25 @@
...
     "prplmesh_docker": {
-        "name": "dockerized_device",
+        "name": "controller",
         "board_type": "prplmesh_docker",
         "role": "controller",
         "conn_cmd": "",
         "devices": [
             {
-                "name": "repeater1",
+                "name": "lan",
```
Comment (arnout) [inline tests/boardfarm_plugins/boardfarm_prplmesh/prplmesh_config.json:9]: Why rename this? And as usual:
- should have been a separate commit;
- should be explained why in the commit message.

(You don't need to split it in a separate commit now, but please do explain in the commit message.)

If the rename is not needed, I'd much prefer it.

File: tests/boardfarm_plugins/boardfarm_prplmesh/tests/client_association_link_metrics.py [hunk 6]
```diff
@@ -19,16 +20,20 @@ class ClientAssociationLinkMetrics(PrplMeshBaseTest):
...
+        # Locate test participants
+        controller = self.dev.DUT.controller_entity
+        agent = self.dev.lan.agent_entity
+        sta = self.dev.wifi
+
+        # This test doesn't work for real HW.
+        # Skip it for prplWRT device.
+        # TODO: Remove as soon, as test works for prplWRT device
+        if self.dev.DUT is PrplMeshPrplWRT:
+            self.skipTest("This test isn't ready for prplWRT devices")
```
Comment (arnout) [inline tests/boardfarm_plugins/boardfarm_prplmesh/tests/client_association_link_metrics.py:32]: Could we instead do something like
```
if not self.dev.has_wifi:
    self.skipTest(...)
```

Or maybe:

```
try:
    sta = self.dev.wifi
except AttributeError:
    self.skipTest(...)
```

Note that the latter could be factored in a setup function. Let's say that PrplMeshBaseTest has a setup method that looks a bit like this:
```
def setUp(self):
    try:
        self.configure()
    except AttributeError, e:
        self.skipTest(... use e to find which device was missing ...)
```

Alternatively, we could add a get_device function that raises the skipTest if needed.

File: tests/boardfarm_plugins/boardfarm_prplmesh/testsuites.cfg [hunk 7]
```diff
@@ -6,3 +6,4 @@
 [test_flows]
 InitialApConfig
 ApConfigRenew
+ClientAssociationLinkMetrics
```
Comment (arnout) [inline tests/boardfarm_plugins/boardfarm_prplmesh/testsuites.cfg:9]: This should have been in the first commit.

**Review by arnout — APPROVED** [review]
I really like how you approached this PR: first commit that is basically copy and paste of the test_flows code, and then a follow-up commit that "fixes" it to work with the boardfarm device. Good example for @aciliketcap.

File: tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py (hunk 4 above)
Comment (abelog) [reply tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py:36]: Yep, it is. `self.name` should be one of boardfarm build-in [enum](https://github.com/mattsm/boardfarm/blob/100521fde1fb67536682cafecc2f91a6e2e8a6f8/boardfarm/lib/DeviceManager.py#L25). If it doesn't match - we can't address devices directly by name in test.

File: tests/boardfarm_plugins/boardfarm_prplmesh/tests/client_association_link_metrics.py (hunk 1 above)
Comment (abelog) [reply tests/boardfarm_plugins/boardfarm_prplmesh/tests/client_association_link_metrics.py]: Actually, part of setup is refactored in #1486

File: tests/boardfarm_plugins/boardfarm_prplmesh/tests/client_association_link_metrics.py (hunk 3 above)
Comment (abelog) [reply tests/boardfarm_plugins/boardfarm_prplmesh/tests/client_association_link_metrics.py:43]: 30 second were chosen blindly.

File: tests/boardfarm_plugins/boardfarm_prplmesh/prplmesh_config.json (hunk 5 above)
Comment (abelog) [reply tests/boardfarm_plugins/boardfarm_prplmesh/prplmesh_config.json:9]: Without rename, we can't address devices by name.

File: tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py (hunk 4 above)
Comment (arnout) [reply tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py:36]: In that case, we should probably (eventually) rename ALEntity.name.

File: tests/boardfarm_plugins/boardfarm_prplmesh/testsuites.cfg (hunk 7 above)
Comment (aciliketcap) [reply tests/boardfarm_plugins/boardfarm_prplmesh/testsuites.cfg:9]: I guess he added it last because test is only working once last commit is in place.

**Review by aciliketcap — APPROVED** [review]

**Review by aciliketcap — APPROVED** [review]
